### PR TITLE
Allow auto termination of spans

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Implements auto-termination of spans. This is achieved by checking whether a span has `AutoTerminationMode.ON_BACKGROUND` just before the session payload is constructed, and terminating it and its descendants if this is true.

## Testing

Added integration tests.

### CODE REVIEW

File: embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/spans/SpanRepository.kt [hunk 1]
```diff
@@ -101,4 +102,50 @@ class SpanRepository {
     }
 
     private fun notTracked(spanId: String): Boolean = activeSpans[spanId] == null && completedSpans[spanId] == null
```
Comment (bidetofevil) [inline embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/spans/SpanRepository.kt:104]: Will you be adding unit tests that verify this works with different edge-case combinations of descendants which might change the traversal logic? Like if a child is no longer active, but the grand child is active?

**Review by bidetofevil — APPROVED** [review]
LGTM. Question about adding more test cases for different combinations of span states. We need to make sure that after backgrounding, all active spans belonging to traces whose root as auto terminate configured will be terminated, irrespective of the state of its ancestors.

**Comment by fractalwrench** [comment]
@bidetofevil I added some extra test cases to the integration test & made some slight tweaks to the implementation, so putting this out for review again.

**Review by bidetofevil — APPROVED** [review]
LGTM
